Adding a part of the `DataGridView` test app to check DGV in Virtual mode

vladimir-krestov opened the pull request:
Just adding a new UI test case to have an additional example

## Proposed changes

- Add a `DataGridView` in Virtual mode with some data as a new test app to `WinformsControlsTest` project
- Rename and refactor the existing UI test app of a normal DGV

<!-- We are in TELL-MODE the following section must be completed -->

## Customer Impact

- Users can see an example of how to implement and set up a `DataGridView` in Virtual mode in their applications

## Regression? 

- No

## Risk

- Minimal

<!-- end TELL-MODE -->


## Screenshots <!-- Remove this section if PR does not change UI -->
- The new test app looks like that:
![image](https://user-images.githubusercontent.com/49272759/100308967-b7dbf400-2fba-11eb-8ee9-0b62ed3703a7.png)




## Test methodology <!-- How did you ensure quality? -->

- UI testing

## Test environment(s) <!-- Remove any that don't apply -->

- Microsoft Windows [Version 10.0.19042.630]
- .NET SDK Version:   6.0.100-alpha.1.20554.10


<!-- Mention language, UI scaling, or anything else that might be relevant -->


###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/4249)

merriemcgaw commented:
@dreddy-work  can you review and merge - this is a result of @RussKie and me asking for Virtual Mode testing for DGV when we added it for ListView.

Tanya-Solyanik commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/DataGridViewHeaders.cs:
```diff
@@ -41,6 +58,23 @@ protected override void OnLoad(EventArgs e)
             currentDPILabel1.Text = DeviceDpi.ToString();
+
+            // Connect the virtual-mode events to event handlers
+            dataGridView2.CellValueNeeded += new DataGridViewCellValueEventHandler(dataGridView2_CellValueNeeded);
```
We don't need the explicit `new`, these even hook-ups can be simplified like this - 
dataGridView2.CellValueNeeded += dataGridView2_CellValueNeeded;

vladimir-krestov replied:
Yes, you are right, but the Designer returns this wrapper every time. I added all changes according to the Designer behavior

Tanya-Solyanik replied:
This is not a part of the designer.cs file, it will not be overridden when you save this form in the designer, designer generates .Net 1.1 code, this is a bug in code-gen. There is no reason to propagate hard to read code.

vladimir-krestov replied:
You are right, I have removed the extra code
Thank you!

Tanya-Solyanik commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/DataGridViewHeaders.cs:
```diff
@@ -22,6 +22,23 @@ public partial class DataGridViewHeaders : Form
+        // Set this value to false to use cell-level commit scope
+        private bool _rowScopeCommit = true;
+
         public DataGridViewHeaders()
```
The file name and the class name no longer reflect what this form is testing. It used to test a HDPI change related to DGV headers 3-4 years ago... Not it's re-purposed to test DGV scenarios, it makes sense to rename the file now.

vladimir-krestov replied:
Renamed to "DataGridViewTest"

Tanya-Solyanik commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/DataGridViewHeaders.cs:
```diff
@@ -89,5 +123,188 @@ private void numericUpDown1_ValueChanged(object sender, EventArgs e)
+        {
+            // If this is the row for new records, no values are needed
+            if (e.RowIndex == dataGridView2.RowCount - 1)
+                return;
```
nit: please add braces around the `return` statement.

vladimir-krestov replied:
Thanks, good point, 
I agree. Actually, I have taken this example from https://docs.microsoft.com/dotnet/desktop/winforms/controls/implementing-virtual-mode-wf-datagridview-control
It has little inconsistencies with our style

Tanya-Solyanik replied:
This is an ancient sample. It would be nice to update it eventually, but not now.

Tanya-Solyanik commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/DataGridViewHeaders.cs:
```diff
@@ -89,5 +123,188 @@ private void numericUpDown1_ValueChanged(object sender, EventArgs e)
+            }
+        }
+
+        private void dataGridView2_CellValuePushed(object sender, System.Windows.Forms.DataGridViewCellValueEventArgs e)
```
Do you need the namespace name in the args type?

vladimir-krestov replied:
No, thanks. We already have a `System.Windows.Forms` using definition.

Tanya-Solyanik commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/DataGridViewHeaders.cs:
```diff
@@ -89,5 +123,188 @@ private void numericUpDown1_ValueChanged(object sender, EventArgs e)
+            if (e.RowIndex < _customers.Count)
+            {
+                // If the user is editing a new row, create a new TestCustomer object
+                _customerInEdit ??= new TestCustomer(_customers[e.RowIndex].Name,
```
please factor out `_customers[e.RowIndex]` into a local variable

Tanya-Solyanik commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/DataGridViewHeaders.cs:
```diff
@@ -89,5 +123,188 @@ private void numericUpDown1_ValueChanged(object sender, EventArgs e)
+
+        private void dataGridView2_CellValuePushed(object sender, System.Windows.Forms.DataGridViewCellValueEventArgs e)
+        {
+            TestCustomer customerTmp = null;
```
Why is this variable needed?

vladimir-krestov replied:
This variable was needed to store a link to a current `TestCustomer` instance and set its properties from a changed line in DGV. 
But this is too confusing and not obvious. 
So I changed the implementation to use `_customerInEdit` field clearly.

Tanya-Solyanik commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/DataGridViewHeaders.cs:
```diff
@@ -89,5 +123,188 @@ private void numericUpDown1_ValueChanged(object sender, EventArgs e)
+        }
+    }
+
+    internal class TestCustomer
```
Please factor out this class into a new file.

Tanya-Solyanik reviewed (commented):
I have some clean-up comments

Tanya-Solyanik commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/DataGridViewTest.Designer.cs:
```diff
@@ -203,5 +269,11 @@ private void InitializeComponent()
         private System.Windows.Forms.Button button2;
+        private System.Windows.Forms.Label label2;
+        private System.Windows.Forms.DataGridView dataGridView2;
+        private System.Windows.Forms.DataGridViewTextBoxColumn PersonName;
```
private fields should not be capitalized

vladimir-krestov commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/DataGridViewTest.cs:
```diff
@@ -0,0 +1,272 @@
+{
+    public partial class DataGridViewTest : Form
+    {
+        private static readonly Font[] s_fonts = new Font[]
```
Made it like this:
![image](https://user-images.githubusercontent.com/49272759/99867437-7d451680-2bca-11eb-8df3-5e1dd04f9b10.png)

vladimir-krestov commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/DataGridViewTest.cs:
```diff
@@ -0,0 +1,272 @@
+            dataGridView1.Rows[2].Cells[1].Value = s_fonts[_rowHeaderFontIndex];
+        }
+
+        private void button2_Click(object sender, EventArgs e)
```
Named as `changeFontButton` and `resetFontButton`

vladimir-krestov commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/DataGridViewTest.cs:
```diff
@@ -0,0 +1,272 @@
+        private static readonly Font[] s_fonts = new Font[]
+            {
+                new Font("Tahoma", 12F, FontStyle.Regular),
+                new Font("Fira Code", 10F, FontStyle.Bold),
```
Just removed

vladimir-krestov commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/TestCustomer.cs:
```diff
@@ -0,0 +1,34 @@
+    ///  It is implementation of a simple customer class
+    ///  to represent the test data items used by code example in <see cref="DataGridViewTest"/>
+    /// </summary>
+    internal class TestCustomer
```
Thank you!
Should I move this class to `IntegrationTests.Common` near by `Person` class? Or maybe to a different folder?

Tanya-Solyanik replied:
Maybe later, when you use this data source elsewhere.

vladimir-krestov commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/DataGridViewTest.cs:
```diff
@@ -0,0 +1,272 @@
+                // If the user is editing a new row, create a new TestCustomer object
+                TestCustomer customerTmp = _customers[e.RowIndex];
+                _customerInEdit ??= new TestCustomer(customerTmp.Name,
+                                                     customerTmp.Age,
```
Sorry, I don't catch what do you mean. It looks Ok to me. 
But I just move all arguments to one line, it looks better for me. This line is 118 characters long so it is readable.

Tanya-Solyanik replied:
```
                _customerInEdit ??= 
                    new TestCustomer(customerTmp.Name,
                    customerTmp.Age,
```
one line will work too

vladimir-krestov commented:
As @Tanya-Solyanik asked, I refactored the existing UI test app for DGV and renamed it. Also, fixed review points for the new app.
As @RussKie asked, I implemented the DGV Virtual mode test case as a separate test app. It is necessary to have a clear example for users how they can implement DGV in Virtual mode

RussKie commented on src/System.Windows.Forms/tests/IntegrationTests/System.Windows.Forms.IntegrationTests.Common/MainFormControlsTabOrder.cs:
```diff
@@ -30,5 +30,6 @@ public enum MainFormControlsTabOrder
         FormBorderStylesButton,
         ToggleIconButton,
         FileDialogButton,
+        DataGridViewInVirtualModeButton
```
Please move it up to line:18

vladimir-krestov replied:
Do you want to see "DataGridViewInVirtualModeTest" button next to "DataGridViewTest" button on the "Main" window?

RussKie replied:
Yes

vladimir-krestov replied:
Fixed. Now it looks like this:
![image](https://user-images.githubusercontent.com/49272759/100314844-9b46b880-2fc8-11eb-97f1-201dcb00ed05.png)

RussKie commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/DataGridViewTest.cs:
```diff
@@ -8,21 +8,17 @@
 namespace WinformsControlsTest
 {
-    public partial class DataGridViewHeaders : Form
+    public partial class DataGridViewTest : Form
```
Do you think this is necessary? This sample app is primarily testing header-related behaviours, isn't it?

vladimir-krestov replied:
I followed this advice: https://github.com/dotnet/winforms/pull/4249#discussion_r526314267

RussKie reviewed (commented):
👍

RussKie commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/DataGridViewInVirtualModeTest.cs:
```diff
@@ -0,0 +1,199 @@
+            _customers.Add(new TestCustomer("Sarah", 67, false, "Female"));
+
+            // Set the row count, including the row for new records
+            dataGridView1.RowCount = 4;
```
Can't we infer this instead of hardcoding?

vladimir-krestov replied:
Thanks, changed it to depending on `_customers` count

RussKie commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/DataGridViewInVirtualModeTest.cs:
```diff
@@ -0,0 +1,199 @@
+        // The example was taken from: https://docs.microsoft.com/dotnet/desktop/winforms/controls/implementing-virtual-mode-wf-datagridview-control
+
+        // Declare a List to serve as the test data source
+        private List<TestCustomer> _customers = new List<TestCustomer>();
```
```suggestion
        private readonly List<TestCustomer> _customers = new List<TestCustomer>();
```

RussKie reviewed (commented):
👍 two minor points
